File: src/features/typing/data/ts/find_median_from_data_stream.ts
class MedianFinder {
    private sorted: number[] = [];

    addNum(num: number): void {
        let lo = 0;
        let hi = this.sorted.length;
        while (lo < hi) {
            const mid = lo + Math.floor((hi - lo) / 2);
            if (this.sorted[mid] < num) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        this.sorted.splice(lo, 0, num);
    }

    findMedian(): number {
        const n = this.sorted.length;
        if (n % 2 === 1) {
            return this.sorted[Math.floor(n / 2)];
        }
        return (this.sorted[n / 2 - 1] + this.sorted[n / 2]) / 2;
    }
}

File: src/features/typing/data/dart/find_median_from_data_stream.dart
class MedianFinder {
  final List<int> _sorted = [];

  void addNum(int num) {
    int lo = 0;
    int hi = _sorted.length;
    while (lo < hi) {
      final int mid = lo + (hi - lo) ~/ 2;
      if (_sorted[mid] < num) {
        lo = mid + 1;
      } else {
        hi = mid;
      }
    }
    _sorted.insert(lo, num);
  }

  double findMedian() {
    final int n = _sorted.length;
    if (n % 2 == 1) {
      return _sorted[n ~/ 2].toDouble();
    }
    return (_sorted[n ~/ 2 - 1] + _sorted[n ~/ 2]) / 2;
  }
}
